File: python/24.py
import unittest
from typing import Optional

from list import ListNode, buildList, compareList

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        first = head
        second = head.next
        first.next = self.swapPairs(second.next)
        second.next = first

        return second


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertTrue(compareList(solution.swapPairs(
            buildList([1, 2, 3, 4])), buildList([2, 1, 4, 3])))

File: typescript/24.ts
import { ListNode, buildList, compareList } from "./list";

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
    if (!head || !head.next) return head;

    const dummyHead = new ListNode(undefined, head);
    let prev = dummyHead;

    while (prev.next && prev.next.next) {
        const first = prev.next;
        const second = prev.next.next;

        first.next = second.next;
        second.next = first;
        prev.next = second;

        prev = first;
    }

    return dummyHead.next;
}

test('24', () => {
    expect(compareList(swapPairs(buildList([1, 2, 3, 4])), buildList([2, 1, 4, 3]))).toBeTruthy();
});